### PR TITLE
Remove typedef ts lint override

### PR DESCRIPTION
**Pull request by leddie24**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ npm run change`

#### Description of changes

Removes typedef tslint override.

**There are a lot of typings that I'm not sure about; I will comment them in the files, let me know if you have a recommendation for a better typing**

#### Focus areas to test

(optional)

### CODE REVIEW

**Comment by micahgodbolt** [comment]
There's a ton of this that appears to be prettier changes. Is there a way we can do 1 prettier pass on all files and make this PR just about the typedef changes?

**Comment by leddie24** [comment]
Yeah, not sure how to turn off the autoformatting; I have the prettier plugin disabled.  I'll try to revert as many inadvertent changes as possible.

**Comment by leddie24** [comment]
@micahgodbolt , all the formatting changes should be reverted now.

File: packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx [hunk 1]
```diff
@@ -521,7 +521,8 @@ describe('SpinButton', () => {
     const exampleDefaultValue = '12';
 
     function delay(millisecond: number): Promise<string> {
-      return new Promise<string>(resolve => setTimeout(resolve, millisecond));
+      // tslint:disable-next-line:no-any
+      return new Promise<string>((resolve: any) => setTimeout(resolve, millisecond));
```
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx]: Any suggestions here? I'm not sure if resolves have a type

File: packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx [hunk 2]
```diff
@@ -600,7 +601,7 @@ describe('SpinButton', () => {
 
   it(`should pass KeyCode ${KeyCodes.enter} to onValidate handler`, () => {
     let keyCode;
-    const onValidate: jest.Mock = jest.fn((value, event) => {
+    const onValidate: jest.Mock = jest.fn((value: any, event: any) => {
```
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx]: Should these be any, or should we disable typedef for *.test files?

File: packages/office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx [hunk 3]
```diff
@@ -5,7 +5,13 @@ import { FocusZone } from 'office-ui-fabric-react/lib/FocusZone';
 import { Image } from 'office-ui-fabric-react/lib/Image';
 import './FocusZone.Photos.Example.scss';
 
-const PHOTOS = createArray(25, () => {
+interface IPhoto {
```
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx]: Is this OK to add an interface here, or should is there an easier/better way to resolve this typedef?

File: common/changes/office-ui-fabric-react/edwl-2018-09-fixTypedefTsLint_2018-09-21-16-26.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Remove typedef tslint override",
+      "type": "minor"
```
Comment (leddie24) [inline common/changes/office-ui-fabric-react/edwl-2018-09-fixTypedefTsLint_2018-09-21-16-26.json]: Should this be `none` or `minor`?

Comment (dzearing) [reply common/changes/office-ui-fabric-react/edwl-2018-09-fixTypedefTsLint_2018-09-21-16-26.json]: I'd definitely recommend minor at the least.

File: packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx (hunk 2 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx]: I would be totally ok with disabling for *.test files.

Maybe even for *.example as well. The typings kind of pollute the message for the example.

File: packages/office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx (hunk 3 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx]: Ok with disabling for tests and examples.

File: common/changes/office-ui-fabric-react/edwl-2018-09-fixTypedefTsLint_2018-09-21-16-26.json (hunk 4 above)
Comment (dzearing) [reply common/changes/office-ui-fabric-react/edwl-2018-09-fixTypedefTsLint_2018-09-21-16-26.json]: I'd keep at minor. There could be unintended breaks here, better be safe.

File: packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx (hunk 1 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.test.tsx]: if possible special case the tslint validation to exclude .test and .example.

**Comment by JasonGore** [comment]
I'm not sure I fully agree with this rule change, mostly because of an issue with tslint's `typedef` rules not working nicely with functional types. The `typedef` rules will require all functional params and return values to have typedefs, even if TS typing is fully enforced with a functional type.

Code such as this functional type, which correctly applies strongly typed props and return values, will now generate a tslint error with this change checked in:

```
export const collapsibleSectionStyles: ICollapsibleSectionComponent['styles'] = (props) => {
```

This will break my createComponent approach, requiring this code to have duplicated types in the declaration, such as:

```
export const collapsibleSectionStyles: ICollapsibleSectionComponent['styles'] = (props: ICollapsibleSectionStyleProps): ICollapsibleSectionStyles => {
```

which is incredibly redundant.

More info:
* https://github.com/palantir/tslint/issues/2654
* https://github.com/palantir/tslint/issues/3524

These changes overall may be good, but I think until the compatibility issues with functional declarations get resolved, we should not actually enable the `typedef` rule in force (or more finely tune it to allow functional types to work without the need for redundant typing.)

**Comment by dzearing** [comment]
This is accruing some merge conflicts and build failures. Is there a way we can disable the rule for tests and examples?

**Comment by dzearing** [comment]
As this has become stale, closing until @leddie24 is back and can address the remaining issues.

**Pull request closed without merging**